frontend/javascripts/oxalis/model/accessors/dataset_accessor.js

```diff
@@ -27,6 +27,123 @@ import { aggregateBoundingBox } from "libs/utils";
 import { formatExtentWithLength, formatNumberToLength } from "libs/format_utils";
 import messages from "messages";
 
+export type ResolutionsMap = Map<number, Vector3>;
+
+export class ResolutionInfo {
+  resolutions: Array<Vector3>;
+  resolutionMap: Map<number, Vector3>;
+
+  constructor(resolutions: Array<Vector3>) {
+    this.resolutions = resolutions;
+    this._buildResolutionMap();
+  }
+
+  _buildResolutionMap() {
+    // Each resolution entry can be characterized by it's greatest resolution dimension.
+    // E.g., the resolution array [[1, 1, 1], [2, 2, 1], [4, 4, 2]] defines that
+    // a zoomstep of 2 corresponds to the resolution [2, 2, 1] (and not [4, 4, 2]).
+    // Therefore, the largest dim for each resolution has to be unique across all resolutions.
+
+    // This function creates a map which maps from powerOfTwo (2**index) to resolution.
+
+    const resolutions = this.resolutions;
+
+    if (resolutions.length !== _.uniqBy(resolutions.map(_.max)).length) {
+      throw new Error("Max dimension in resolutions is not unique.");
+    }
+
+    this.resolutionMap = new Map();
+    for (const resolution of resolutions) {
+      this.resolutionMap.set(_.max(resolution), resolution);
+    }
+  }
+
+  getResolutionsWithIndices(): Array<[number, Vector3]> {
+    return Array.from(this.resolutionMap.entries()).map(entry => {
+      const [powerOfTwo, resolution] = entry;
+      const resolutionIndex = Math.log2(powerOfTwo);
+      return [resolutionIndex, resolution];
+    });
+  }
+
+  indexToPowerOf2(index: number): number {
+    return 2 ** index;
+  }
+
+  hasIndex(index: number): boolean {
+    const powerOfTwo = this.indexToPowerOf2(index);
+    return this.resolutionMap.has(powerOfTwo);
+  }
+
+  getResolutionByIndex(index: number): ?Vector3 {
+    const powerOfTwo = this.indexToPowerOf2(index);
+    return this.getResolutionByPowerOf2(powerOfTwo);
+  }
+
+  getResolutionByIndexOrThrow(index: number): Vector3 {
+    const resolution = this.getResolutionByIndex(index);
+    if (!resolution) {
+      throw new Error(`Resolution with in index ${index} does not exist`);
+    }
+    return resolution;
+  }
+
+  getResolutionByIndexWithFallback(index: number): Vector3 {
+    const resolutionMaybe = this.getResolutionByIndex(index);
+    if (resolutionMaybe) {
+      return resolutionMaybe;
+    } else {
+      const powerOf2 = this.indexToPowerOf2(index);
+      return [powerOf2, powerOf2, powerOf2];
```

Comment:
> This fallback doesn't have anything to do with the fallback rendering, right?

Correct, the reason for this code is that the previous code always assumed that all layers have all magnifications (trusting that missing mags would be caught later on). Out of caution, I implemented this fallback here, since I didn't want to check all the callers for their compatibility with the new assumption.

However, I just noticed that this fallback is not really correct, since it always assumes an isotropic magnification. The old code solved this by actually falling back to the magnifications of other layers, however, this information is not really available here. I could add it here, but I'll try to work around this somehow else probably. Thanks for raising attention to this here :+1:

Reply:
Hm, I tried to get rid of the with-fallback code but couldn't do it completely. See my latest commit here: https://github.com/scalableminds/webknossos/pull/4755/commits/74aa5fd13a4b3c09505060da66e3876faf73d494

One major use case is having a global position and then transforming that position to a given mag. If mag 1 is missing, there definitely needs to be a fallback somewhere. For mag 1, this is easy, since it is not ambiguous (all components of [1,1,1] have to be 1, otherwise it wouldn't be mag 1). But for other mags, one would actually need the mag information of other layers.
I'm afraid the entire resolution code still needs some refactoring. The `ResolutionInfo` class helps already a bit, but it's still not perfect (as the above issue shows). Also, `zoomedAddressToAnotherZoomStep` is used in quite a lot of places. I added this topic to https://github.com/scalableminds/webknossos/issues/4838.